vllm/model_executor/models/bert.py

```diff
@@ -573,6 +573,281 @@ def _decode_token_type_ids(input_ids: torch.Tensor) -> torch.Tensor:
+        self.pooling = pooling
+        self.remove_cls_sep = remove_cls_sep
+
+    def get_supported_tasks(self) -> Set[PoolingTask]:
+        return {"embed"}
+
+    def get_pooling_updates(self, task: PoolingTask) -> PoolingParamsUpdate:
+        return PoolingParamsUpdate(requires_token_ids=True)
+
+    def forward(
+        self,
+        hidden_states: Union[torch.Tensor, list[torch.Tensor]],
+        pooling_metadata: PoolingMetadata,
+    ) -> Union[torch.Tensor, list[torch.Tensor]]:
+        if isinstance(hidden_states, torch.Tensor):
+            hs_list = [hidden_states]
+        else:
+            hs_list = list(hidden_states)
+
+        for i, hs in enumerate(hs_list):
+            if hs.dim() == 3 and hs.size(0) == 1:
+                hs_list[i] = hs.squeeze(0)  # [L, H]
+            elif hs.dim() != 2:
+                raise ValueError(f"Expected [L,H] or [1,L,H], got {tuple(hs.shape)}")
+
+        B = len(hs_list)
+        H = hs_list[0].size(-1)
+
+        raw_lens = pooling_metadata.prompt_lens
+
+        def _fallback_lens_from_hs():
+            return [int(h.size(0)) for h in hs_list]
+
+        if raw_lens is None:
+            lens = _fallback_lens_from_hs()
+        elif isinstance(raw_lens, int):
+            lens = [int(raw_lens)] * B
+        else:
+            try:
+                tmp = list(raw_lens)
+                if len(tmp) == B:
+                    lens = [int(x) for x in tmp]
+                elif len(tmp) == 1:
+                    lens = [int(tmp[0])] * B
+                else:
+                    lens = _fallback_lens_from_hs()
+            except TypeError:
+                lens = _fallback_lens_from_hs()
+
+        max_len = max(int(h.size(0)) for h in hs_list)
+        device = hs_list[0].device
+
+        # pad to [B, T, H]
+        padded = hs_list[0].new_zeros((B, max_len, H))  # zeros
+        attn_mask = torch.zeros((B, max_len), dtype=torch.bool, device=device)
+
+        for i, (hs, L) in enumerate(zip(hs_list, lens)):
+            L = int(L)
+            L = min(L, max_len)
+            padded[i, :L] = hs[:L]
+            attn_mask[i, :L] = True
+
+        # [CLS]/[SEP] remove(Optional)
+        token_ids = getattr(pooling_metadata, "prompt_token_ids", None)  # [B,T] or None
+        if self.remove_cls_sep and token_ids is not None:
+            for i, L in enumerate(lens):
+                L = int(min(L, max_len))
+                if L <= 0:
+                    continue
+                if (
+                    self.cls_token_id is not None
+                    and int(token_ids[i, 0].item()) == self.cls_token_id
+                ):
+                    attn_mask[i, 0] = False
+                if (
+                    self.sep_token_id is not None
+                    and int(token_ids[i, L - 1].item()) == self.sep_token_id
+                ):
+                    attn_mask[i, L - 1] = False
+
+        # MLM logits: [B, T, V]
+        B, T, _ = padded.shape
```

Comment:
If `padded` was initialized to have B as the first dimension, then there is no need to redefine B as the size of the first dimension. And `T` is `max_len`, right?